### PR TITLE
Wrap SDK initialization API in a timeout

### PR DESCRIPTION
**Pull request by MarcinJuraszek**
Addressed an issue reported in https://github.com/OfficeDev/microsoft-teams-library-js/issues/719. I don't really know what the correct timeout period would be. There isn't much logic happening on the host side before we respond to the `initialize` call, so we could probably make it even shorter, but I want to be conservative to prevent false positives.

At some point we should look into a way to reject the promise if we're not running in a context of Teams/Office/Outlook without having to wait for the timeout.

### CODE REVIEW

**Comment by erinha** [comment]
<p>Analyzed commit id : 3c14339a9be5b3b58e878ab074df90685dbfdb81</p><hr><details><summary><b><span style="color: coral">⯅</span> @microsoft/bundle-analysis-app:</b> +277 Bytes</summary><table><tr><th>Metric Name</th><th>Baseline Size</th><th>Compare Size</th><th>Size Diff</th></tr><tr><td>main.js</td><td>57.34 KB</td><td>57.61 KB</td><td><span style="color: coral">⯅</span> +277 Bytes</td></tr><tr><td>Total Size</td><td>57.34 KB</td><td>57.61 KB</td><td><span style="color: coral">⯅</span> +277 Bytes</td></tr></table></details><hr><p>Baseline commit: 8103a864bee9c57df538093125f1707df748e936</p>

File: packages/teams-js/src/internal/utils.ts [hunk 1]
```diff
@@ -266,3 +266,31 @@ export function callCallbackWithErrorOrResultOrNullFromPromiseAndReturnPromise<T
   });
   return p;
 }
+
+/**
+ * A helper function to add a timeout to an asynchronous operation.
+ *
+ * @param action Action to wrap the timeout around
+ * @param timeoutInMs Timeout period
+ * @param timeoutError Error to reject the promise with if timeout elapses before the action completed
```
Comment (aeunms) [inline packages/teams-js/src/internal/utils.ts:275]: Nit: Maybe we could tag @internal like the other functions in the file

**Comment by erinha** [comment]
<p>Analyzed commit id : 9cc221abb6441653259fc36276e689b4cecb3dce</p><hr><details><summary><b><span style="color: coral">⯅</span> @microsoft/bundle-analysis-app:</b> +277 Bytes</summary><table><tr><th>Metric Name</th><th>Baseline Size</th><th>Compare Size</th><th>Size Diff</th></tr><tr><td>main.js</td><td>57.34 KB</td><td>57.61 KB</td><td><span style="color: coral">⯅</span> +277 Bytes</td></tr><tr><td>Total Size</td><td>57.34 KB</td><td>57.61 KB</td><td><span style="color: coral">⯅</span> +277 Bytes</td></tr></table></details><hr><p>Baseline commit: 8103a864bee9c57df538093125f1707df748e936</p>

File: packages/teams-js/src/internal/utils.ts [hunk 2]
```diff
@@ -266,3 +266,33 @@ export function callCallbackWithErrorOrResultOrNullFromPromiseAndReturnPromise<T
...
+ *
+ * @internal
+ */
+export function runWithTimeout<TResult, TError>(
+  action: () => Promise<TResult>,
+  timeoutInMs: number,
+  timeoutError: TError,
+): Promise<TResult> {
+  return new Promise((resolve, reject) => {
+    const timeoutHandle = window.setTimeout(reject, timeoutInMs, timeoutError);
```
Comment (aeunms) [inline packages/teams-js/src/internal/utils.ts:287]: I don't know a ton about promises or timeout functions so this is a learning opportunity for me- From reading the docs on [setTimeout()](https://developer.mozilla.org/en-US/docs/Web/API/setTimeout), the 'timeoutError' argument seems to be passed into 'action', but the example we see here doesn't seem to use the argument and I can't find elsewhere that timeoutError gets used. Did I understand what's going on correctly? If so, did you want to use timeoutError somewhere?

Comment (MarcinJuraszek) [reply packages/teams-js/src/internal/utils.ts:287]: Everything after the second parameter will be pushed as arguments to the first parameter when the timer fires.

```
window.setTimeout(reject, timeoutInMs, timeoutError);
``` 

is basically equivalent to

```
window.setTimeout(() => reject(timeoutError), timeoutInMs)
```

but it's saving a closure allocation and level of indirection with the new lambda.

Comment (aeunms) [reply packages/teams-js/src/internal/utils.ts:287]: Awesome, thank you :)

**Review by aeunms — DISMISSED** [review]
LGTM!

**Comment by erinha** [comment]
<p>Analyzed commit id : 726b0a6496493405a4a360e0d20ed38582560720</p><hr><details><summary><b><span style="color: coral">⯅</span> @microsoft/bundle-analysis-app:</b> +277 Bytes</summary><table><tr><th>Metric Name</th><th>Baseline Size</th><th>Compare Size</th><th>Size Diff</th></tr><tr><td>main.js</td><td>57.34 KB</td><td>57.61 KB</td><td><span style="color: coral">⯅</span> +277 Bytes</td></tr><tr><td>Total Size</td><td>57.34 KB</td><td>57.61 KB</td><td><span style="color: coral">⯅</span> +277 Bytes</td></tr></table></details><hr><p>Baseline commit: d4fca89a93d4ee7dc26ada0b2613893953712853</p>

**Comment by erinha** [comment]
<p>Analyzed commit id : 8cc6573b842048b1469b34c4082581f60535f91e</p><hr><details><summary><b><span style="color: coral">⯅</span> @microsoft/bundle-analysis-app:</b> +277 Bytes</summary><table><tr><th>Metric Name</th><th>Baseline Size</th><th>Compare Size</th><th>Size Diff</th></tr><tr><td>main.js</td><td>57.34 KB</td><td>57.61 KB</td><td><span style="color: coral">⯅</span> +277 Bytes</td></tr><tr><td>Total Size</td><td>57.34 KB</td><td>57.61 KB</td><td><span style="color: coral">⯅</span> +277 Bytes</td></tr></table></details><hr><p>Baseline commit: 2a662cc0e2f2fcfb61cf9865896d162e019c04de</p>

**Comment by erinha** [comment]
<p>Analyzed commit id : f22e75d9d9a943ea017dd6f646ef96915fe878b8</p><hr><details><summary><b><span style="color: coral">⯅</span> @microsoft/bundle-analysis-app:</b> +277 Bytes</summary><table><tr><th>Metric Name</th><th>Baseline Size</th><th>Compare Size</th><th>Size Diff</th></tr><tr><td>main.js</td><td>57.34 KB</td><td>57.61 KB</td><td><span style="color: coral">⯅</span> +277 Bytes</td></tr><tr><td>Total Size</td><td>57.34 KB</td><td>57.61 KB</td><td><span style="color: coral">⯅</span> +277 Bytes</td></tr></table></details><hr><p>Baseline commit: 3196b790d5553bc86f87085d791ec42c2eb53d66</p>

**Comment by erinha** [comment]
<p>Analyzed commit id : 20d5e77b47483fe4cfbce01592f13464509c85dc</p><hr><details><summary><b><span style="color: coral">⯅</span> @microsoft/bundle-analysis-app:</b> +1.4 KB</summary><table><tr><th>Metric Name</th><th>Baseline Size</th><th>Compare Size</th><th>Size Diff</th></tr><tr><td>main.js</td><td>57.27 KB</td><td>58.67 KB</td><td><span style="color: coral">⯅</span> +1.4 KB</td></tr><tr><td>Total Size</td><td>57.27 KB</td><td>58.67 KB</td><td><span style="color: coral">⯅</span> +1.4 KB</td></tr></table></details><hr><p>Baseline commit: 1eb3b020461fc39c8808f1f8153097e002cbc4bd</p>

**Comment by erinha** [comment]
$(bundleAnalysisTask.bundleAnalysisComment)

File: packages/teams-js/src/internal/utils.ts [hunk 3]
```diff
@@ -266,3 +266,33 @@ export function callCallbackWithErrorOrResultOrNullFromPromiseAndReturnPromise<T
   });
   return p;
 }
+
+/**
+ * A helper function to add a timeout to an asynchronous operation.
+ *
+ * @param action Action to wrap the timeout around
+ * @param timeoutInMs Timeout period
```
Comment (AE-MS) [inline packages/teams-js/src/internal/utils.ts:274]: >Timeout period [](http://example.com/codeflow?start=22&length=14)

I know it's literally in the name of the variable but perhaps consider adding "in milliseconds" to this description

**Comment by AE-MS** [comment]
>    * but after the frame is loaded successfully.

Should this new timeout behavior be documented? I don't know if we should think of this change as "doing this to behave better when not running inside an host but we don't really want to tell people to do that regularly" or we should think of it as "yes, we recommend you do this to determine whether or not you're running inside a host"

---
Refers to: packages/teams-js/src/public/app.ts:433 in 06af57c. [](commit_id = 06af57ccc7998e868d043c64b2e2eab0b8807162, deletion_comment = False)

**Review by AE-MS — APPROVED** [review]
:shipit:

File: packages/teams-js/src/internal/utils.ts (hunk 2 above)
Comment (AE-MS) [inline packages/teams-js/src/internal/utils.ts:287]: >window [](http://example.com/codeflow?start=26&length=6)

Can we use `setTimeout` without explicitly referring to `window` ?

**Review by AE-MS — CHANGES_REQUESTED** [review]
:clock1:

**Comment by erinha** [comment]
<p>Analyzed commit id : 27b0d73c5f6cdb9416c741af3d877957661a0840</p><hr><details><summary><b><span style="color: coral">⯅</span> @microsoft/bundle-analysis-app:</b> +255 Bytes</summary><table><tr><th>Metric Name</th><th>Baseline Size</th><th>Compare Size</th><th>Size Diff</th></tr><tr><td>main.js</td><td>58.94 KB</td><td>59.19 KB</td><td><span style="color: coral">⯅</span> +255 Bytes</td></tr><tr><td>Total Size</td><td>58.94 KB</td><td>59.19 KB</td><td><span style="color: coral">⯅</span> +255 Bytes</td></tr></table></details><hr><p>Baseline commit: f33919b6b97b4d1836bce2dc8f5f50d8b0190c29</p>

**Review by AE-MS — APPROVED** [review]
:shipit:

**Review by erinha — APPROVED** [review]

**Pull request merged**